### PR TITLE
Refactor K:Unblockable > Static

### PR DESCRIPTION
**Pull request by Northmoc**

### CODE REVIEW

**Comment by Hanmac** [comment]
"CARDNAME can't block." might need to be refactored too if able

**Comment by Northmoc** [comment]
> "CARDNAME can't block." might need to be refactored too if able

A few of these did include something like that: https://github.com/Card-Forge/forge/pull/1551/files#diff-694431d534d1d47f75f2289bb6c5bb6b563284c2be8247b07c9b44eab64992ef

Think we should add the others to this PR or do a different one?

**Comment by Northmoc** [comment]
> i think we can make "CARDNAME can't block." in another one and merge this if it is ready

Yeah, this one will be big enough

**Comment by Hanmac** [comment]
testEquippedAbilities TestCase needs to be refactored, should probably use a different thing and not something with "Unblockable"

File: forge-ai/src/main/java/forge/ai/CreatureEvaluator.java [hunk 1]
```diff
@@ -59,9 +59,9 @@ public int evaluateCreature(final Card c, final boolean considerPT, final boolea
         }
         if (c.hasKeyword(Keyword.HORSEMANSHIP)) {
             value += addValue(power * 10, "horses");
-        }
-        if (c.hasKeyword("Unblockable")) {
-            value += addValue(power * 10, "unblockable");
+        //} TODO is there a way to parse the CantBlockBy static here?
```
Comment (Agetian) [inline forge-ai/src/main/java/forge/ai/CreatureEvaluator.java:62]: I guess you'd need to loop over all the static effects on the card and look for CantBlockBy and inspect its parameters to determine that it's the "unblockable" effect. Might not be the most performance-efficient way of doing it though, but I don't know a better way :/ Maybe someone else would have a better idea?

File: forge-ai/src/main/java/forge/ai/ability/AttachAi.java [hunk 2]
```diff
@@ -661,9 +661,9 @@ private static Card attachAICuriosityPreference(final SpellAbility sa, final Lis
             if (card.hasKeyword(Keyword.HORSEMANSHIP)) {
                 cardPriority += 40;
             }
-            if (card.hasKeyword("Unblockable")) {
-                cardPriority += 50;
-            }
+            //if (card.hasKeyword("Unblockable")) { pseudo "keyword" refactored to static – how to refactor this?
```
Comment (Agetian) [inline forge-ai/src/main/java/forge/ai/ability/AttachAi.java:664]: Same thoughts and concerns as above here :)

**Comment by Hanmac** [comment]
the AI stuff might be touched in a later MR

**Comment by Agetian** [comment]
I would argue that the AI here needs to be dealt with without leaving it for later (which is often never), as merging as is will make the AI considerably stupider as it'll miss a lot of offensive opportunities otherwise when it can deal unblockable damage ;)

**Comment by Hanmac** [comment]
there is another unblockable token for the fishes:
https://github.com/Card-Forge/forge/blob/master/forge-gui/res/tokenscripts/u_1_1_fish_unblockable.txt

**Comment by Agetian** [comment]
One of the ways I can see this might work for the AI is something like this:

```
            for (StaticAbility st : tgtCard.getStaticAbilities()) {
                if (st.getParam("Mode").equals("CantBlockBy") && "Creature.Self".equals(st.getParam("ValidAttacker"))) {
                    // modify the evaluation score for the card
                    break; // stop evaluating if we found what we're looking for
                }
            }

```

Not sure how this might need to be extended to be more "specific" when needed (and if needed), but hopefully this can be used as a base for the change. Yes, it iterates over all the static abilities, but at least it cuts the loop short if the Unblockable ability is found.

**Review by Agetian — APPROVED** [review]

**Pull request merged**